### DISCUSSION TITLE
Design of cursor visibility and cursor position API

### DISCUSSION THREAD

**Discussion by proneon267** [beeware/toga#1885] [Ideas]
As per our discussion until now, here is what I think the API should look like:

**Window by Window Basis:**
```
# As a getter:
print(window.cursor_position)
# Returns the current cursor position relative to the specified window as a position tuple (x,y).

print(window.cursor_visible)
# Returns the current cursor visibility for the specified window as a Boolean.

# As a setter:
window.cursor_position = (x, y)
# Sets the cursor position relative to the specified window.

window.cursor_visible = True/False
# Sets the cursor visibility for the specified window.
```

**App wide Basis:**
```
# As a getter:
print(app.cursor_position)
# Returns the current absolute cursor position as a position tuple (x,y).

print(app.cursor_visible)
# Returns the current cursor visibility for the whole app as a Boolean.

# As a setter:
app.cursor_position = (x, y)
# Sets the absolute cursor position.

app.cursor_visible = True/False
# Sets the cursor visibility for the whole app.
```

Here, I can see the usefulness of exposure of `cusor_visibility` and `cursor_position` both as window basis and app wide basis.

**`cursor_visibility` on a window basis** can allow the users to interact with the other windows of the app while preventing interaction on a particular window. 

**`cursor_visibility` on an app wide basis** can allow the developers to prevent user interaction for the whole app. This is useful when the app doesn't want the user to do something while a process or task is running which would be hampered by any user interaction.

**`cursor_position` on a window basis** can be used to implement a "first time app use guide". This would allow the developer to move the cursor to a particular position to point out a feature on the app. This would be useful as the app developer would want to move the cursor with respect to particular window on which the feature is located.

**`cursor_position` on an app wide basis** can be used to point out certain things on the user's screen which is not related to the app itself, like a feature on the user's system.

@freakboy3742 Let me know how do you feel about this.

**Comment by freakboy3742** [comment]
First off - I'll flag that broadly speaking, I'm *not* a huge fan of the general idea of cursor control APIs. From a UX perspective, apps should keep their hands off the position of the cursor; and the only time an app should care about the *position* of a cursor is (a) the position at the time of a click event on a widget like a canvas, and (b) enter/exit region events.

Broad cursor visibility is also needed for use cases like full screen presentation apps - but outside that sort of case, apps should *not* be making the cursor arbitrarily invisible. The cursor is a key UX element. Making that element invisible is, generally speaking, a *bad* UX move.

Continuing specific discussion points from #1864:

1. If we were to add a cursor position API - what is the use case for a *window* level API for setting the position of a cursor? There is only one cursor. The cursor is not window specific. Why is the *position* of the cursor a window-specific feature? I don't understand how "first time app use guide" is an answer to this; and `app.cursor_position = (window.position.x + 10, window.position.y + 20)` would seem to address the "window relative position" question. 
2. What is the use case for *window* level API for cursor visibility. Again, the making the cursor disappear is, generally speaking, a *bad* idea from a UX perspective. 

I've also had an in-person discussion with @mhsmith; and he made the good point that cursor visibility is arguably a *style* question. In CSS, [cursor](https://developer.mozilla.org/en-US/docs/Web/CSS/cursor) is a property that can control both *which* cursor is used when hovering over a widget, as well as hiding the cursor on a per-widget basis.

In terms of the specific API proposal - you've covered the public interface API; but there's no mention of the backend API. However it's probably better to wait until we have a concrete public interface before we work out how the backends will implement that API.

**Comment by proneon267** [comment]
The true use of cursor position system would be to detect hover over a widget and do some operations during that event, like it's done in CSS. 

Addressing the questions:

1.Yes, there is only 1 cursor but there will be multiple windows:
Suppose there are 2 windows in an app: `window1` and `window2`. Now, there is a widget on `window1`. 
So, the developer wants to point the cursor to this widget, so he can do:
```
window1.cursor_position = window.widget.position
```
While you could do:
```
app.cursor_position = (window1.position.x + widget.position.y, window1.position.y + widget.position.y)
```
But, clearly the second approach is more verbose.
Window based cursor positioning system will take the burden off the developer to manually calculating the appropriate position. Moreover, it will improve readability of the code.

This is not a new idea. Most of the toolkits provide both the window relative and global cursor positioning API. Maybe I am not the most qualified one to suggest use cases for this API but clearly there is a need for such API, hence they exist in most toolkits.

2.Apps contain windows. Apps are not windows. Hiding the cursor app wide will be a bad UX decision as the user won't be able to differentiate between a non responding app and a app that doesn't want to take user interaction.

cursor visibility on an window basis provides more fine grained control to the developer. They can choose to show/hide cursor on a specific window or hide entirely for all the windows of their app.

Use case for window level cursor visibility would be an app having multiple windows and the developer wants to hide the cursor on a specific window, rather than app wide.

**Reply by freakboy3742** [reply]
> The true use of cursor position system would be to detect hover over a widget and do some operations during that event, like it's done in CSS.

The CSS approach is nothing like what you've described in this API. CSS is a lot closer to the "widget enter/exit event" approach. It doesn't provide a "get the current position of the cursor", or "set the current position of the cursor". 

> But, clearly the second approach is more verbose. Window based cursor positioning system will take the burden of the developer to manually calculating the appropriate position. Moreover, it will improve readability of the code.

Sure - but once again, that doesn't answer the question "why are they doing this in the first place". Taking control of the cursor is *bad* UX. An incredibly flexible and convenient API for something developers *shouldn't* be doing in the first place isn't an especially desirable proposition.
 
> This is not a new idea. Most of the toolkits provide both the window relative and global cursor positioning API. Maybe I am not the most qualified one to suggest use cases for this API but clearly there is a need for such API, hence they exist in most toolkits.

"Everyone else does it" isn't an inherently strong argument. Every GUI toolkit has an API for building menus. Toga doesn't, for a *very* specific UX reason. 

> Use case for window level cursor visibility would be an app having multiple windows and the developer wants to hide the cursor on a specific window, rather than app wide.

Again - "a window level API will allow people to control visibility on a window-by-window basis" is inherently obvious. It doesn't address *why* this is a useful or desirable feature, or why it this isn't a *style*-based feature.

**Reply by proneon267** [reply]
We can do a "widget enter/exit event" approach. This would provide the most closet API for what the developer would want to achieve as their end goal.

I am fully in favor of a "widget enter/exit event" approach. What do you have in mind for a such an API?

**Reply by freakboy3742** [reply]
I would have thought the API was fairly obvious - an on_enter event handler and an on_exit event handler.

However, I'm *not* in favour of adding such an API. On the list of features that I'd like to see added to Toga, this is pretty far down the list - and it would come with a not insignificant performance hit, as moving the cursor across the screen would fire dozens of events.

More importantly, you haven't responded to the direct question that I've asked: *Why* do you want to do this? What is the use case? What *actual problem* are you having that you need to solve? What is the reason for needing this functionality? "So people can show and hide the cursor on a per-window basis" or "so people can move the cursor" aren't answers to this question. 

I've given you the use case for hiding the cursor at the level of the app - presentation tools. The only rationale you've given for proposing these extra APIs is "so it can be done". The fact that it *can* be done doesn't mean it *should* be done.

**Reply by proneon267** [reply]
My personal use case is this:
![]( https://user-images.githubusercontent.com/45512885/232501265-4cd03f9e-1834-4a01-a5aa-b1dcd2e117ea.png )
I want to hide the cursor for the middle window and keep them visible for the side windows. Only, window by window basis would allow this to happen.

**Reply by freakboy3742** [reply]
You've posted that image before - but you haven't answered *why* you want to hide the cursor. I'd argue it would make your app massively confusing. How do you know where the cursor is *if you can't see it*? How do you know where to move it so you *can* see it again?

**Reply by proneon267** [reply]
If hiding the cursor in a window would confuse the user then wouldn't hiding the cursor app wide would also be equally confusing?

If hiding the cursor in presentation mode is the use case then why can't we allow an app developer to make a presentation mode that is not full screened and still hide the cusor on that window?

**Reply by freakboy3742** [reply]
Yes - hiding the cursor *is* confusing. That's why it should almost *never* be done. I can and have pointed at a very specific use case where it *is* appropriate - slide show apps - where the app is put into a read-only, full screen display context, where control isn't being done by mouse. 

The reason you can't do "presentation mode" not full screened is that you still need to interact with the window. When you're in a full screen presentation mode, there's literally no purpose to having a mouse, because there's nothing to click on. Non-full screen, you have access to window decorations and other windows in other apps - and you need to know where the mouse is to interact with those.

The fact that full screen and cursor visibility are separate APIs means the two are decoupled, and you could make the cursor invisible without being full screen; I'm not sure there's any good way to avoid that. However, we don't need to needlessly complexify the API to support *both* window and app cursor visibility - and we can avoid issues like answering "is the cursor visible" at both the app level and the window level. If the cursor is invisible at the level of one specific window... is it visible at the app level? We could try to synthesise an answer to that question... or we can avoid the question entirely by only having cursor visibility as an app-level concern.

**Reply by proneon267** [reply]
I want to take the example of a presentation app:

If I enter full screen mode then there would be 2 full screened windows: the slide deck screen and the speaker's note screen. 

I would want to hide the cursor on the presentation screen to prevent distraction to the viewers. But, I would like to keep the cursor on the speaker's note screen (Maybe because I use the cursor to keep track of the sentence which I am saying😅). 

Now, if there will be only app level cursor visibility, how would I achieve the above in my app?

**Reply by freakboy3742** [reply]
Well - there's only one cursor. If the cursor is on your speakers notes, it's not on the slide deck :-) There's no need to make the cursor invisible at all - and if the cursor *was* invisible on the slide deck, you'd have no way to work out where it was so that you *could* move it to the speakers notes.

**Reply by proneon267** [reply]
I agree with you :-) . That seems logical and would work in the app. Should we make cursor visibility an app wide API only? Also, what about cursor position? Thinking about it a bit more, it seems like there isn't a very good use case for it. The only use case would be a "on_enter/on_exit widget" API where the cursor position would be used internally. Should we also just leave out the cursor position API too?

**Reply by freakboy3742** [reply]
I'm pretty sure I've been saying consistently that cursor visibility should be an app-wide API only... so... yes - cursor visibility should be an app-wide API only. 

I'm also very sceptical whether a set position API is needed at all. Any event being captured that was based on cursor position will either be provided the cursor position as part of the event (e.g., a mouse move event on a canvas), or doesn't require an explicit position because the mouse being in the right place is implicit in the event firing in the first place (e.g., on-enter/exit).

**Reply by proneon267** [reply]
I agree with you. Well then, is the PR needed at all? There was already, `show_cursor()` and `hide_cursor()`. Is another property `cursor_visible` needed? If not, should I close that PR and implement the backend in terms of the already present API?

**Reply by freakboy3742** [reply]
That would seem to be the right approach to me. 

(and this is why a design discussion is valuable :-)

**Reply by proneon267** [reply]
Thanks for helping me :)
